Improting the Dependencies

In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [4]:
#loading the dataset to a panda dataframe
credit_card_data = pd.read_csv('/content/creditcard.csv')

In [7]:
#print first five from data set
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [9]:
#print last 5 data sets
credit_card_data.tail()

Time        V1        V2        V3        V4        V5        V6  \
140698  83876  1.297031  0.260234 -0.165836  0.304460  0.238056 -0.084983   
140699  83876  1.200272  0.259410  0.397862  0.631576 -0.476806 -0.882324   
140700  83877 -0.939078  1.234767  1.433774 -0.097576 -0.255769 -0.689204   
140701  83879  1.170286  0.563313 -0.271928  1.394477 -0.038260 -1.000654   
140702  83879 -1.376888  0.304186  2.232899 -0.740000       NaN       NaN   

              V7        V8        V9  ...       V21       V22       V23  \
140698 -0.056115  0.029952 -0.115999  ... -0.315545 -0.948937 -0.024322   
140699 -0.062362 -0.028223 -0.130967  ... -0.231443 -0.739076  0.151545   
140700  0.509961  0.153369  0.179511  ... -0.257956 -0.582912  0.080816   
140701  0.126524 -0.031662  0.139095  ... -0.038141 -0.086158 -0.033513   
140702       NaN       NaN       NaN  ...       NaN       NaN       NaN   

             V24       V25       V26       V27       V28  Amount  Class  
140698 -0.883643  0.335230  0.150161 -0.029612  0.008220    0.99    0.0  
140699  0.475489  0.134602  0.066882 -0.031207  0.020989    0.89    0.0  
140700  0.334513 -0.257496  0.094942  0.487942  0.239537    4.49    0.0  
140701  0.220516  0.517620 -0.312217  0.041102  0.053760    1.00    0.0  
140702       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [10]:
#check the information of data sets
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140703 entries, 0 to 140702
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    140703 non-null  int64  
 1   V1      140703 non-null  float64
 2   V2      140703 non-null  float64
 3   V3      140703 non-null  float64
 4   V4      140703 non-null  float64
 5   V5      140702 non-null  float64
 6   V6      140702 non-null  float64
 7   V7      140702 non-null  float64
 8   V8      140702 non-null  float64
 9   V9      140702 non-null  float64
 10  V10     140702 non-null  float64
 11  V11     140702 non-null  float64
 12  V12     140702 non-null  float64
 13  V13     140702 non-null  float64
 14  V14     140702 non-null  float64
 15  V15     140702 non-null  float64
 16  V16     140702 non-null  float64
 17  V17     140702 non-null  float64
 18  V18     140702 non-null  float64
 19  V19     140702 non-null  float64
 20  V20     140702 non-null  float64
 21  V21     14

In [12]:
#check if missing value is present or not
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        1
V6        1
V7        1
V8        1
V9        1
V10       1
V11       1
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [13]:
# distribution of legit and fraud transactions
credit_card_data['Class'].value_counts()

0.0    140438
1.0       264
Name: Class, dtype: int64

140438 is legit transaction and 264 is fraud transaction here. dikh jata hai saf saf with vlue. count wale function se

In [14]:
#separating the data for better analysis
legit= credit_card_data[credit_card_data.Class==0]
fraud= credit_card_data[credit_card_data.Class==1]

print(legit.shape)
print(fraud.shape)

(140438, 31)
(264, 31)


In [15]:
#getting the statistical measures of the data that is present in amount
legit.Amount.describe()

count    140438.000000
mean         91.068217
std         247.065709
min           0.000000
25%           6.000000
50%          23.900000
75%          80.760000
max       19656.530000
Name: Amount, dtype: float64

In [17]:
fraud.Amount.describe()

count     264.000000
mean      115.387008
std       245.189867
min         0.000000
25%         1.000000
50%         9.560000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [18]:
#compare the values for the transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    52034.876842 -0.239504  0.011245  0.687392  0.131415 -0.275448   
1.0    42420.212121 -5.610819  3.946507 -7.143814  4.512206 -3.977333   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.082153 -0.106135  0.062394 -0.088075  ...  0.042180 -0.041927   
1.0   -1.480669 -5.936405  1.508791 -2.604923  ...  0.239899  1.256981   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.118336 -0.033151  0.012195  0.130022  0.023506 -0.000321  0.002165   
1.0   -0.311690 -0.116543 -0.107463  0.197988  0.056528  0.489643  0.081783   

           Amount  
Class              
0.0     91.068217  
1.0    115.387008  

[2 rows x 30 columns]

method of under sampling

build a sample dataset containig the similar distribution of normal and fraud transaction

In [19]:
legit_sample= legit.sample(n=264)

concatenating two data frames

In [25]:
new_dataset=pd.concat([legit_sample, fraud],axis=0)

In [26]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
42207   40997  1.227712 -0.136154 -0.729066  0.142166  1.885531  3.784708   
32268   36718  1.298745 -0.753397  0.250113 -0.725015 -1.106515 -0.572771   
84566   60366 -2.050232 -0.406106  0.421793  1.972699  1.352423 -2.121695   
91287   63397 -0.401171 -0.819427  2.352016 -0.936208 -0.842698 -0.759768   
126442  78006  1.326361  0.339201 -0.112349  0.455655  0.134660 -0.492602   

              V7        V8        V9  ...       V21       V22       V23  \
42207  -0.724894  0.965960  0.440649  ... -0.391312 -1.131745  0.069912   
32268  -0.634357  0.028230 -0.813847  ... -0.456022 -1.055702  0.079080   
84566  -0.366461  0.384869 -0.700313  ...  0.221834 -0.026228 -0.232753   
91287  -0.697916 -0.256655 -2.496402  ... -0.304723 -0.507804  0.038506   
126442  0.078311 -0.149859  0.073792  ... -0.351391 -1.000612  0.002293   

             V24       V25       V26       V27       V28  Amount  Class  
42207   0.964262  0.526972 -0.625065  0.048630  0.026792   12.99    0.0  
32268  -0.037138  0.035334  0.909458 -0.077820  0.000256   44.73    0.0  
84566   0.608286  0.055515 -0.261823  0.066282 -0.200492    1.00    0.0  
91287   0.341594 -0.161445 -0.153544 -0.057888 -0.118785   25.90    0.0  
126442 -0.657960  0.356429  0.156173 -0.023293  0.021093    2.28    0.0  

[5 rows x 31 columns]

In [22]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
125342  77627 -7.139060  2.773082 -6.757845  4.446456 -5.464428 -1.713401   
128479  78725 -4.312479  1.886476 -2.338634 -0.475243 -1.185444 -2.112079   
131272  79540 -0.114361  1.036129  1.984405  3.128243 -0.740344  1.548619   
135718  81372 -0.885254  1.790649 -0.945149  3.853433 -1.543510  0.188582   
137705  82289 -1.464897  1.975528 -1.077145  2.819191  0.069850 -0.789044   

              V7        V8        V9  ...       V21       V22       V23  \
125342 -6.485365  3.409395 -3.053493  ...  1.303250 -0.016118 -0.876670   
128479 -2.122793  0.272565  0.290273  ...  0.550541 -0.067870 -1.114692   
131272 -1.701284 -2.203842 -1.242265  ... -1.032935  1.196428 -0.112857   
135718 -2.988383  1.344059 -2.294535  ...  0.746160  0.550802 -0.034882   
137705 -1.196101  0.673654 -1.363724  ...  0.174099 -0.272505 -0.031549   

             V24       V25       V26       V27       V28  Amount  Class  
125342  0.382230 -1.054624 -0.614606 -0.766848  0.409424  106.90    1.0  
128479  0.269069 -0.020572 -0.963489 -0.918888  0.001454   60.00    1.0  
131272  0.254719  0.696668  0.482370  0.129969  0.223924    0.20    1.0  
135718 -0.567608 -0.528318  0.258782  0.506893  0.176736    7.57    1.0  
137705 -0.406166  0.157769 -0.104393  0.073796 -0.041570    1.00    1.0  

[5 rows x 31 columns]

In [27]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    51891.128788 -0.328771 -0.199625  0.515531  0.281927 -0.354574   
1.0    42420.212121 -5.610819  3.946507 -7.143814  4.512206 -3.977333   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.111807 -0.159021  0.060867 -0.039513  ... -0.072411 -0.022128   
1.0   -1.480669 -5.936405  1.508791 -2.604923  ...  0.239899  1.256981   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.123398  0.006152  0.015143  0.181513  0.004471  0.065938 -0.006591   
1.0   -0.311690 -0.116543 -0.107463  0.197988  0.056528  0.489643  0.081783   

           Amount  
Class              
0.0    103.360568  
1.0    115.387008  

[2 rows x 30 columns]

In [30]:
X= new_dataset.drop('Class',axis=1)
Y=new_dataset['Class']

In [31]:
print(X)

         Time        V1        V2        V3        V4        V5        V6  \
42207   40997  1.227712 -0.136154 -0.729066  0.142166  1.885531  3.784708   
32268   36718  1.298745 -0.753397  0.250113 -0.725015 -1.106515 -0.572771   
84566   60366 -2.050232 -0.406106  0.421793  1.972699  1.352423 -2.121695   
91287   63397 -0.401171 -0.819427  2.352016 -0.936208 -0.842698 -0.759768   
126442  78006  1.326361  0.339201 -0.112349  0.455655  0.134660 -0.492602   
...       ...       ...       ...       ...       ...       ...       ...   
125342  77627 -7.139060  2.773082 -6.757845  4.446456 -5.464428 -1.713401   
128479  78725 -4.312479  1.886476 -2.338634 -0.475243 -1.185444 -2.112079   
131272  79540 -0.114361  1.036129  1.984405  3.128243 -0.740344  1.548619   
135718  81372 -0.885254  1.790649 -0.945149  3.853433 -1.543510  0.188582   
137705  82289 -1.464897  1.975528 -1.077145  2.819191  0.069850 -0.789044   

              V7        V8        V9  ...       V20       V21       V22  \


In [32]:
print(Y)

42207     0.0
32268     0.0
84566     0.0
91287     0.0
126442    0.0
         ... 
125342    1.0
128479    1.0
131272    1.0
135718    1.0
137705    1.0
Name: Class, Length: 528, dtype: float64


In [34]:
#splitting data into training data and testing data
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2,stratify=Y, random_state=2)

In [35]:
print(X.shape, X_train.shape, X_test.shape)

(528, 30) (422, 30) (106, 30)


Model Training
Logistic Reression

In [42]:
model = LogisticRegression()

In [43]:
#Trainig the Logistic regression model with Training data
model.fit(X_train,Y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [44]:
#accuracy on training data
X_train_prediction= model.predict(X_train)
training_data_accuracy=accuracy_score(X_train_prediction, Y_train)

In [45]:
print('Accuracy on train data :', training_data_accuracy)

Accuracy on train data : 0.943127962085308


In [46]:
#accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy=accuracy_score(X_test_prediction, Y_test)

In [47]:
print('Accuracy score on test data :', test_data_accuracy)

Accuracy score on test data : 0.9528301886792453
